pymc_experimental/utils/linear_cg.py

```diff
@@ -0,0 +1,279 @@
+    rhs = rhs / rhs_norm
+
+    # residuals
+    residual = rhs - mat @ initial_guess
+    batch_shape = residual.shape[:-2]
+
+    result = np.copy(initial_guess)
+
+    if not np.allclose(residual, residual):
+        raise RuntimeError("NaNs encountered when trying to perform matrix-vector multiplication")
+
+    # sometimes we are lucky and preconditioner solves the system right away
+    # check for convergence
+    residual_norm = np.linalg.norm(residual, 2)
+    has_converged = residual_norm < stop_updating_after
+
+    if has_converged.all() and not n_tridiag:
+        n_iter = 0  # skip iterations
+    else:
+        precond_residual = preconditioner(residual)
+
+        curr_conjugate_vec = precond_residual
+        residual_inner_prod = residual.T @ precond_residual
+
+        # define storage matrices
+        mul_storage = np.zeros_like(residual)
+        alpha = np.zeros((*batch_shape, 1, rhs.shape[-1]))
+        beta = np.zeros_like(alpha)
+        is_zero = np.zeros((*batch_shape, 1, rhs.shape[-1]))
+
+    # Define tridiagonal matrices if applicable
+    if n_tridiag:
+        t_mat = np.zeros((n_tridiag_iter, n_tridiag_iter, *batch_shape, n_tridiag))
+        alpha_tridiag_is_zero = np.zeros(*batch_shape, n_tridiag)
+        alpha_reciprocal = np.zeros(*batch_shape, n_tridiag)
+        prev_alpha_reciprocal = np.zeros_like(alpha_reciprocal)
+        prev_beta = np.zeros_like(alpha_reciprocal)
+
+    update_tridiag = True
+    last_tridiag_iter = 0
+
+    # it is possible that we don't reach tolerance even after all the iterations are over
+    tolerance_reached = False
+
+    # start iteration
+    for k in range(n_iter):
+        mvms = mat @ curr_conjugate_vec
+        if precond:
+            alpha = curr_conjugate_vec @ mvms  # scalar
+
+            # safe division
+            is_zero = alpha < eps
+            alpha = masked_fill(alpha, mask=is_zero, fill_value=1)
+            alpha = residual_inner_prod / alpha
+            alpha = masked_fill(alpha, mask=is_zero, fill_value=0)
+
+            # cancel out updates by setting directions which have converged to zero
+            alpha = masked_fill(alpha, mask=has_converged, fill_value=0)
+            residual = residual - alpha * mvms
+
+            # update precond_residual
+            precond_residual = preconditioner(residual)
+
+            # Everything inside _jit_linear_cg_updates
+            linear_cg_retvals = linear_cg_updates(
+                result,
+                alpha,
+                residual_inner_prod,
+                eps,
+                beta,
+                residual,
+                precond_residual,
+                curr_conjugate_vec,
+            )
+
+            (
+                result,
+                alpha,
+                residual_inner_prod,
+                eps,
+                beta,
+                residual,
+                precond_residual,
+                curr_conjugate_vec,
+            ) = linear_cg_retvals
```

Comment:
This is just a small nitpick, but you could combine this, so you don't need to create the `linear_cg_retvals`.

Reply:
Fixed in the latest commit